src/expr/core/src/sig/udf.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! UDF implementation interface.
+//!
+//! To support a new language or runtime for UDF, implement the interface in this module.
+//!
+//! See expr/impl/src/udf for the implementations.
+
+use anyhow::{Context, Result};
+use arrow_array::RecordBatch;
+use futures::stream::BoxStream;
+use risingwave_common::types::DataType;
+
+/// The global registry of UDF implementations.
+///
+/// To register a new UDF implementation:
+///
+/// ```ignore
+/// #[linkme::distributed_slice(UDF_IMPLS)]
+/// static MY_UDF_LANGUAGE: UdfImplDescriptor = UdfImplDescriptor {...};
+/// ```
+#[linkme::distributed_slice]
+pub static UDF_IMPLS: [UdfImplDescriptor];
+
+/// Find a UDF implementation by language.
+pub fn find_udf_impl(
+    language: &str,
+    runtime: Option<&str>,
+    link: Option<&str>,
+) -> Result<&'static UdfImplDescriptor> {
+    UDF_IMPLS
+        .iter()
+        .find(|desc| (desc.match_fn)(language, runtime, link))
```

Comment:
Shall we also check whether there are multiple implementations `match` the given metadata, just in case of confusion or indeterminism?